src/Swift.Bindings/src/Marshaler/BoundGenericsHandler.cs

```diff
@@ -0,0 +1,146 @@
+    /// Translates the Swift generic type of the given property declaration into a C# type name.
+    /// </summary>
+    /// <param name="propertyDecl">The property declaration.</param>
+    /// <returns>The C# type name with generic parameters.</returns>
+    /// <exception cref="NotSupportedException">Thrown when the property is not bound generic.</exception>
+    public string TranslateToCSharpName(PropertyDecl propertyDecl)
+    {
+        if (!IsBoundGeneric(propertyDecl))
+            throw new NotSupportedException(
+                $"Attempted to translate to C# name for a non-bound generic property {propertyDecl.Name}");
+        var namedTypeSpec = (NamedTypeSpec)propertyDecl.SwiftTypeSpec;
+        return TranslateToCSharpName(namedTypeSpec);
+    }
+
+    /// <summary>
+    /// Translates the Swift generic type of the given argument declaration into a C# type name.
+    /// </summary>
+    /// <param name="argumentDecl">The argument declaration.</param>
+    /// <returns>The C# type name with generic parameters.</returns>
+    /// <exception cref="NotSupportedException">Thrown when the argument is not bound generic.</exception>
+    public string TranslateToCSharpName(ArgumentDecl argumentDecl)
+    {
+        if (!IsBoundGeneric(argumentDecl))
+            throw new NotSupportedException(
+                $"Attempted to translate to C# name for a non-bound generic argument {argumentDecl.Name}");
+        var namedTypeSpec = (NamedTypeSpec)argumentDecl.SwiftTypeSpec;
+        return TranslateToCSharpName(namedTypeSpec);
+    }
+
+    /// <summary>
+    /// Helper method to convert a Swift <see cref="NamedTypeSpec"/> into its corresponding C# type name.
+    /// </summary>
+    /// <param name="namedTypeSpec">The named type specification.</param>
+    /// <returns>The C# type name string.</returns>
+    /// <exception cref="NotSupportedException">
+    /// Thrown when any generic parameter is not a named type specification.
+    /// </exception>
+    private string TranslateToCSharpName(NamedTypeSpec namedTypeSpec)
+    {
+        List<string> translatedGenericParameters = new();
+        foreach (var genericParameter in namedTypeSpec.GenericParameters)
+        {
+            if (genericParameter is not NamedTypeSpec namedGenericParameter)
+                throw new NotSupportedException(
+                    $"Generic parameter {genericParameter} is not a named type spec");
+            translatedGenericParameters.Add(TranslateToCSharpName(namedGenericParameter));
+        }
+
+        var typeReference = _typeDatabase.GetTypeRecordOrAnyType(namedTypeSpec); // TODO: consider throwing an exception instead
+        return typeReference.CSTypeIdentifier +
+               (translatedGenericParameters.Count > 0
+                    ? $"<{string.Join(", ", translatedGenericParameters)}>"
+                    : "");
+    }
+
+    /// <summary>
+    /// Gets the buffer type name used for marshalling the specified bound generic argument.
```

Comment:
Could you please explain what the "buffer type" really means here - if it's the type which is used for lowering or something else?
Currently the method will succeed only for 2 types - that feels weird. For hardcoded types it's OK, but what about user types...

Reply:
So if a generic projection is a class we will need some "buffer" struct for the purpose of doing the lowering. There are multiple options of generating those buffers / helpers - if we add a runtime support for lowering generics the buffer could be a generic struct, if not we will need to generate a struct per generic instantiation e.g. `Foo_Int_Int> for `Foo<int,int>`. The idea is described in binding-generics.md

It supports only two types as those are required for milestone one and do not require the things mentioned above. Both the array and the set have layout independent of the type parameter which can be represented as a non-generic struct for lowering purposes.

Reply:
Thanks for the explanation